File: 프로그래머스/0/120861. 캐릭터의 좌표/캐릭터의 좌표.py
def solution(keyinput, board):
    answer = []
    width = (board[0] - 1) // 2
    height = (board[1] - 1) // 2
    
    x, y = 0, 0
    for i in keyinput :
        move_x = 0
        move_y = 0
        if i == "left" :
            move_x = -1
        elif i == "right" :
            move_x = 1
        elif i == "up" :
            move_y = 1 
        elif i == "down" :
            move_y = -1 
        if x + move_x < -width or x + move_x > width or y + move_y < -height or y + move_y > height :
            continue
        else :
            x += move_x
            y += move_y
    answer = [x, y]
    return answer

File: 프로그래머스/0/120861. 캐릭터의 좌표/캐릭터의 좌표.java
import java.util.*;

class Solution {
    public int[] solution(String[] keyinput, int[] board) {
        int[] answer = {0, 0};
        Map<String, Integer[]> map = new HashMap<>(Map.of(
                        "up", new Integer[]{0, 1},
                        "down", new Integer[]{0, -1},
                        "left", new Integer[]{-1, 0},
                        "right", new Integer[]{1, 0}));
        int maxX = (board[0] - 1) / 2;
        int minX = -maxX;
        int maxY = (board[1] - 1) / 2;
        int minY = -maxY;

        for (String key : keyinput) {
            Integer[] move = map.get(key);
            int dx = move[0];
            int dy = move[1];
            if (answer[0] + dx >= minX && answer[0] + dx <= maxX && 
                answer[1] + dy >= minY && answer[1] + dy <= maxY) {
                answer[0] += dx;
                answer[1] += dy;
            }
        }

        return answer;
    }
}